### PR TITLE
Feature/Add Policies to Relationships

### PR DESCRIPTION
**Pull request by bjoernmu**
Signed-off-by: Björn Müller <bjoern.mueller@kanu-baerchen.de>

Add Policies to Relationships in Topology Modeler

- [x] Ensure that you followed http://eclipse.github.io/winery/dev/ToolChain#github---prepare-pull-request.
- [x] Branch name checked. That means, the branch name starts with `thesis/`, `fix/`, ...
- [x] Ensure that the commit message is [a good commit message](https://github.com/joelparkerhenderson/git_commit_message)
- [x] Ensure to use auto format in **all** files
- [x] Ensure that you appear in `NOTICE` at Copyright Holders
- [ ] Tests created for changes
- [ ] Documentation updated (if needed)
- [ ] Screenshots added (for UI changes)

### CODE REVIEW

**Comment by miwurster** [comment]
@bjoernmu @ghareeb-falazi: can you guys check what is left here and check the relation to the `project/radon` branch?

**Comment by miwurster** [comment]
@lharzenetter please revise and merge.

**Comment by miwurster** [comment]
Also, please check relation to https://github.com/OpenTOSCA/winery/tree/feature/policesForRelationshipTemplates

File: org.eclipse.winery.frontends/app/topologymodeler/src/app/services/policy.service.ts [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache Software License 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ *******************************************************************************/
+
+@Injectable()
+export class PolicyService {
+    @Output() addNewPolicyEvent: EventEmitter<any>;
```
Comment (lharzenetter) [inline org.eclipse.winery.frontends/app/topologymodeler/src/app/services/policy.service.ts]: Why do you need this service?
Also, the output annotation can only be used in components.

File: org.eclipse.winery.frontends/app/topologymodeler/src/app/redux/reducers/winery.reducer.ts [hunk 2]
```diff
@@ -286,32 +286,61 @@ export const WineryReducer =
...
-                                        ]
-                                    } : {
-                                        policy: [
-                                            policy
-                                        ]
-                                    }) : nodeTemplate
-                            )
-                    }
-                };
+                if (lastState.currentJsonTopology.nodeTemplates.includes(newPolicy.nodeId)) {
```
Comment (lharzenetter) [inline org.eclipse.winery.frontends/app/topologymodeler/src/app/redux/reducers/winery.reducer.ts]: Why don't you just use a second action that adds policies to a relationship template?

File: org.eclipse.winery.frontends/app/topologymodeler/src/app/canvas/canvas.component.ts [hunk 3]
```diff
@@ -1580,6 +1582,7 @@ export class CanvasComponent implements OnInit, OnDestroy, OnChanges, AfterViewI
             }
         });
         this.entityTypes.relationshipTypes.forEach(value => this.assignRelTypes(value));
+        this.policyService.addNewPolicyEvent.subscribe(data => this.toggleModalHandler(data));
```
Comment (lharzenetter) [inline org.eclipse.winery.frontends/app/topologymodeler/src/app/canvas/canvas.component.ts]: This should be handled by the redux store, if a new policy is added, a corresponding action must be triggered which changes the store. A corresponding subscription detects the changes and changes the UI accordingly.

File: org.eclipse.winery.frontends/app/topologymodeler/src/app/services/policy.service.ts [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0, or the Apache Software License 2.0
+ * which is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+ *******************************************************************************/
```
Comment (lharzenetter) [inline org.eclipse.winery.frontends/app/topologymodeler/src/app/services/policy.service.ts:16]: please move the header to the top of the file

File: org.eclipse.winery.frontends/app/topologymodeler/src/app/canvas/canvas.component.ts [hunk 5]
```diff
@@ -2220,4 +2232,81 @@ export class CanvasComponent implements OnInit, OnDestroy, OnChanges, AfterViewI
         this.layoutDirective.layoutNodes(this.nodeChildrenArray, this.allRelationshipTemplates);
         this.ngRedux.dispatch(this.topologyRendererActions.executeLayout());
     }
+
+    /**
+     * Adds a policy icon to a relationshipTemplate
+     * @param currentRelationships
+     */
+
```
Comment (lharzenetter) [inline org.eclipse.winery.frontends/app/topologymodeler/src/app/canvas/canvas.component.ts:2240]: Please remove the empty line

File: org.eclipse.winery.frontends/app/topologymodeler/src/app/topology-renderer/topology-renderer.module.ts [hunk 6]
```diff
@@ -80,6 +80,7 @@ import { WineryFeatureToggleModule } from '../../../../tosca-management/src/app/
     ],
     exports: [
         TopologyRendererComponent,
+        PoliciesComponent,
```
Comment (lharzenetter) [inline org.eclipse.winery.frontends/app/topologymodeler/src/app/topology-renderer/topology-renderer.module.ts:84]: Please remove the empty line.

**Comment by bjoernmu** [comment]
Follow up at: https://github.com/eclipse/winery/pull/483

**Pull request closed without merging**